### PR TITLE
Fix a panic on invalid source

### PR DESCRIPTION
**Pull request by jvmakine**
Also fixes an issue where running the `main` via IntelliJ ends up in an infinite loop in shell detection.

### CODE REVIEW

File: sources/sources.go [hunk 1]
```diff
@@ -70,13 +70,15 @@ func (s *Sources) Sync(p *ui.UI, force bool) error {
 
 // ForURIs returns Source instances for given uri strings
 func ForURIs(b *ui.UI, dir, env string, uris []string) (*Sources, error) {
-	sources := make([]Source, len(uris))
-	for i, uri := range uris {
+	var sources []Source
```
Comment (lyonlai) [inline sources/sources.go:73]: `make([]Source, 0, len(uris))` maybe? the total length of sources can't exceed length of uris, right?

**Review by lyonlai — APPROVED** [review]
minor comment on array initialisation

File: sources/sources.go (hunk 1 above)
Comment (jvmakine) [reply sources/sources.go:73]: True, will change 👍

**Review by alecthomas — APPROVED** [review]

File: sources/sources.go [hunk 2]
```diff
@@ -70,13 +70,15 @@ func (s *Sources) Sync(p *ui.UI, force bool) error {
...
-	sources := make([]Source, len(uris))
-	for i, uri := range uris {
+	sources := make([]Source, 0, len(uris))
+	for _, uri := range uris {
 		s, err := getSource(b, uri, dir, env)
 		if err != nil {
 			return nil, errors.WithStack(err)
 		}
-		sources[i] = s
+		if s != nil {
```
Comment (alecthomas) [inline sources/sources.go:79]: Shouldn't an invalid source report an error though? Or is it displayed as a warning?

Comment (jvmakine) [reply sources/sources.go:79]: It is displayed as a warning in the getSource function

**Pull request merged**